docs/proposals/volume-provisioning.md

```diff
@@ -125,30 +128,250 @@ We propose that:
+    resources:
+      requests:
+        storage: 4Gi
+  #  volumeName: must be empty!
+  ```
+
+  Example of the created PV:
+
+  ```yaml
+  apiVersion: v1
+  kind: PersistentVolume
+  metadata:
+    annotations:
+      pv.kubernetes.io/provisioned-by: foo.org/foo-volume
+      volume.beta.kubernetes.io/storage-class: myClass
+      foo.org/provisioner: "any other annotations as needed"
+    labels:
+        foo.org/my-label: "any labels as needed"
+    generateName: "foo-volume-"
+  spec:
+    accessModes:
+    - ReadWriteOnce
+    awsElasticBlockStore:
+      fsType: ext4
+      volumeID: aws://us-east-1d/vol-de401a79
+    capacity:
+      storage: 4Gi
+    claimRef:
+      apiVersion: v1
+      kind: PersistentVolumeClaim
+      name: fooclaim
+      namespace: default
+      resourceVersion: "53"
+      uid: 5a294561-7e5b-11e6-a20e-0eb6048532a3
+    persistentVolumeReclaimPolicy: Delete
+  ```
+
+  As result, Kubernetes has a PV that represents the storage asset and is bound
+  to the claim. When everything went well, Kubernetes completed binding of the
+  claim to the PV.
+
+  Kubernetes was not blocked in any way during the provisioning and could
+  either bound the claim to another PV that was created by user or even the
+  claim may have been deleted by the user. In both cases, Kubernetes will mark
+  the PV to be delete using the protocol below.
+
+  The external provisioner MAY save any annotations to the claim that is
+  provisioned, however the claim may be modified or even deleted by the user at
+  any time.
+
+
+### Controller workflow for deleting volumes
+
+When the controller decides that a volume should be deleted it performs these
+steps:
+
+1. The controller changes `pv.Status.Phase` to `Released`.
+
+2. The controller looks for `pv.Annotations["pv.kubernetes.io/provisioned-by"]`.
+   If found, it uses this provisioner/deleter to delete the volume.
+
+3. If the volume is not annotated by `pv.kubernetes.io/provisioned-by`, the
+   controller inspects `pv.Spec` and finds in-tree deleter for the volume.
+
+4. If the deleter found by steps 2. or 3. is internal, it calls it and deletes
+   the storage asset together with the PV that represents it.
+
+5. If the deleter is not known to Kubernetes, it does not do anything.
+
+6. External deleters MUST watch for PV changes. When
+   `pv.Status.Phase == Released && pv.Annotations['pv.kubernetes.io/provisioned-by'] == <deleter name>`,
+   the deleter:
+
+   * It MUST check reclaim policy of the PV and ignore all PVs whose
+     `Spec.PersistentVolumeReclaimPolicy` is not `Delete`.
+
+   * It MUST delete the storage asset.
+
+   * Only after the storage asset was successfully deleted, it MUST delete the
+     PV object in Kubernetes.
+
+   * Any error SHOULD be sent as an event on the PV being deleted and the
+     deleter SHOULD retry to delete the volume periodically.
+
+   * The deleter SHOULD NOT use any information from StorageClass instance
+     referenced by the PV.
```

Comment:
@jsafrane thanks, but then, if there are more than one storage class, the external provisioner need to keep track of the parameters of each SC internally and fetch and use it whenever a delete request comes in. iic on this assumption, it will be difficult to maintain the pointers to SC and PVs when deleter is invoked in the external provisioners . Isnt it ?